### DISCUSSION TITLE
Difference between commands for math mode detection

### DISCUSSION THREAD

**Discussion by TryerGit** [sleepymalc/VSCode-LaTeX-Inkscape#16] [General]
Hello,

From [here](https://github.com/sleepymalc/VSCode-LaTeX-Inkscape/blob/main/VSCode-setting/Snippets/latex.hsnips), math mode detection is done via:

` return  context.scopes.findLastIndex(s => s.startsWith("meta.math")) > context.scopes.findLastIndex(s => s.startsWith("comment") || s.startsWith("meta.text.normal.tex"));`

while on the original Hypersnips website, it is done via more minimal:

`return context.scopes.some(s => s.startsWith("meta.math"));`

Could the difference in the semantics between the two forms be briefly explained? Whatever is the language in which these commands are written (Python ? Javascript?) , I am not familiar with that. Hence this question.

Thank you.

**Comment by sleepymalc** [comment]
I believe this is written in `Javascript`. So basically, the scope can be seen from `Developer: Inspect Editor Tokens and Scopes` in VS Code. And when you write $\LaTeX$, the following might be encountered:
<img width="391" alt="image" src="https://user-images.githubusercontent.com/60427058/211700062-a685ebad-1d33-4df4-90f3-1959c702ed75.png">
<img width="426" alt="image" src="https://user-images.githubusercontent.com/60427058/211700110-cee8b6f3-d245-4179-859c-7edc880926ff.png">
<img width="441" alt="image" src="https://user-images.githubusercontent.com/60427058/211700126-6f648012-97ec-4062-9d91-b36bd666dda7.png">
So if you simply use `s.startsWith("meta.math")`, since all these cases have one entry starting with "meta.math", even if your cursor is in `\[ \text{|} \]`, the math scope is still triggered (which you don't really want since you're trying to type some text, not formula). To handle this nested case is to simply identify the last scope by using `s.findLastIndex()`.

**Comment by TryerGit** [comment]
Hi @sleepymalc 

I am encountering some trouble with your function for math scoping vs the minimal one taken from Hypersnips website. Using the minimal one renders snippets properly while using the one you linked to does not trigger. Do you have any suggestions to help debug why one works and not the other? Can I go incrementally starting from the minimal one? If I can, then, I can try out 3 or 4 incremental functions leading from the minimal one to yours and perhaps that will help figure out where the function fails?

Thanks.

**Reply by sleepymalc** [reply]
You can open the extension log and utilize the functionality of `Developer: Inspect Editor Tokens and Scopes` (by pressing `shift`+`cmd`+`p` and searching for this) to help debug. I should note that, however, Hypersnips sometimes will falsely detect the scope due to some of the VS Code implementation issues; for example, it sometimes will think you're in math mode while you're not (not even in a nested case): see [here](https://github.com/draivin/hsnips/issues/79).

In general, without further information, I don't know how to help.

**Reply by TryerGit** [reply]
Thank you. I worked with your suggestion and created two scenarios.

I have a snippet for the norm like so:

```
context math(context)
snippet nrm "Norm " iA
|| $1 || $0
endsnippet

```

which should be autotriggered in math contexts. 

In the first scenario, with your suggested function, this is not triggered. Then, I copy over the minimal function, save the .hsnips file and retry typing `nrm` and it does expand.

An animated gif is attached which shows these steps. Any help is appreciated. 
![latexworkshop](https://user-images.githubusercontent.com/70434658/211739101-e571d136-fcbd-45a2-b7d0-5f5e1a8c89e2.gif)

**Reply by sleepymalc** [reply]
I'll suggest you investigate the extension log via `Developer: Open Extension Logs Folder`. It's weird since this works on my system. Also opening an issue in the Hypersnips repo may be a good idea.
